Adds the ability to edit the name of the mindmap

JamesMura opened the pull request.

JamesMura commented:
this mostly on covers the happy path , it does little in the way of error handling and validation (checking the title , checking the user owns the map they are editing)

adityamukho commented on pages/api/mindmaps/[key]/index.js:
```diff
@@ -4,6 +4,12 @@ import { rg2cy } from '../../../../utils/cyHelpers'
 
 const { db } = require('../../../../utils/arangoWrapper')
+const updateMindMap = async (database, mindmapKey, name) => {
+  return database.query({
```
This needs to go through RecallGraph's API instead of a direct DB update. That way, RG can track change history for this event.

Please see https://github.com/adityamukho/hivemind/blob/dbdda9325c2e35348ef5716dad280d60e01c8344/pages/api/nodes/index.js#L61 for an example of how to do this,

adityamukho commented on pages/api/mindmaps/[key]/index.js:
```diff
@@ -82,8 +88,15 @@ const MindMapAPI = async (req, res) => {
         }
-
-        return res.status(404).json({ message: 'Not Found.' })
+      case 'POST':
```
A PATCH request would be better to signify this is an update, and also align better with the conventions used elsewhere in this codebase.

adityamukho replied:
Also, this should go into `pages/api/mindmaps/index.js` since:
1. The key is not read from the path,
2. It becomes consistent with the convention used in the _nodes_ API.

adityamukho reviewed (changes requested):
Have suggested some changes. Will deploy to Vercel once the update is routed through RG.

adityamukho commented on pages/api/mindmaps/[key]/index.js:
```diff
@@ -82,8 +88,15 @@ const MindMapAPI = async (req, res) => {
+        if (!name || 0 === name.length) {
+          return res.status(400).json({"message": "name is required"})
+        }
+        await updateMindMap(db, key, name)
```
You can check if the user has write access before calling `updateMindMap`. Please see https://github.com/adityamukho/hivemind/blob/dbdda9325c2e35348ef5716dad280d60e01c8344/pages/api/mindmaps/index.js#L73. In fact, this API endpoint is already equipped to handle name changes with permission checks. You just need to add the validation and then call this endpoint from the frontend code.

adityamukho replied:
After the `isNil` filter is applied to eliminate null/undefined values. You can check if the name field is still present. If present, it should not be empty.

```
if (name && 0 === name.length) {
  return res.status(400).json({"message": "name is required"})
}
```

JamesMura commented:
- Added support for connecting to arrangodb using https
- Added a tooltipped button component 
- Added tooltips for top menu items 
- rename now uses the patch api 
- formatting some of the component classes

adityamukho commented on utils/arangoWrapper.js:
```diff
@@ -1,12 +1,18 @@
+  ARANGO_DB: databaseName,
+  ARANGO_USER: username,
+  ARANGO_PASSWORD: password,
+} = process.env;
```
Object destructuring for `process.env.*` would not work, or would lead to security holes in Next.js, since `process.env` undergoes special handling in order to keep server-side secrets safe and not accidentally bundle them into client-side code. All such vars must therefore be **explicitly** referenced as `process.env.YOUR_CUSTOM_ENV_VAR`

See https://nextjs.org/docs/basic-features/environment-variables

adityamukho commented on db-setup/init.js:
```diff
@@ -3,12 +3,19 @@ const { config } = require('dotenv')
+  ARANGO_DB: databaseName,
+  ARANGO_USER: username,
+  ARANGO_PASSWORD: password,
+} = process.env;
```
OK to destructure here since this code isn't processed by Next.js.

adityamukho commented on components/mindmap/action-items/Fit.js:
```diff
@@ -1,18 +1,37 @@
+import { Maximize } from "react-feather";
+import { Button, Tooltip } from "reactstrap";
+import { runLayout } from "../../../utils/cyHelpers";
+import GlobalContext from "../../GlobalContext";
```
Would've been easier to read if the existing code conventions had been followed (eg. single-quoted string literals, no semicolon at end of statement, etc). Deviating from the convention may not break the code, but it makes the diffs larger and thus more difficult to review.

adityamukho commented on components/mindmap/action-items/ToolTippedButton.js:
```diff
@@ -0,0 +1,23 @@
+      </span>
+
+      <Tooltip
+        placement="top"
```
placement could also be a prop?

adityamukho commented on functions/index.js:
```diff
@@ -1,19 +1,26 @@
+  port,
+  username,
+  password,
+  databaseName,
```
`db: databaseName`

adityamukho commented on functions/index.js:
```diff
@@ -1,19 +1,26 @@
+  protocol = "http",
+  host,
+  port,
+  username,
```
`user: username`

This is required unless also accompanied by updated instructions in `INSTALLATION.md` to change the function config var names. Actually, better to stick with the old names so as to not break existing installations.

adityamukho commented on pages/api/mindmaps/index.js:
```diff
@@ -68,6 +68,9 @@ const MindMapsAPI = async (req, res) => {
         return res.status(response.statusCode).json({ message })
 
       case 'PATCH':
+        if (req.body._id ===  undefined) {
```
If we're checking, better to use _.isEmpty, to catch all types of blanks (null, empty string, undefined, 0, etc).

JamesMura replied:
I think this is perfectly fine as it is now

adityamukho commented on components/mindmap/action-items/Fit.js:
```diff
@@ -1,18 +1,37 @@
-};
+  return (
+    <>
+      <Button
```
Can't use the `TooltippedButton` component here?

adityamukho commented on components/mindmap/action-items/Rename.js:
```diff
@@ -0,0 +1,98 @@
+
+  return (
+    <>
+      <Button className="ml-1" outline color="secondary" id="rename">
```
TTB?

adityamukho commented on components/mindmap/action-items/Search.js:
```diff
@@ -72,6 +73,8 @@ export default function search () {
     <Button className="ml-1" outline color="secondary" id="search" onClick={toggle}>
       <Search size={16}/>
     </Button>
+    <Tooltip placement="top" target="search"  isOpen={tooltipOpen}
```
TTB?

adityamukho commented on components/mindmap/action-items/ShowAll.js:
```diff
@@ -15,8 +15,12 @@ function handler (cyWrapper) {
-  return <Button className="ml-1" outline color="secondary" onClick={() => handler(cyWrapper)}>
+  const [tooltipOpen, setTooltipOpen] = useState(false);
+  return <>
+  <Tooltip placement="top" target="showall"  isOpen={tooltipOpen}
```
TTB?

adityamukho reviewed (changes requested):
Some changes suggested.
